### ISSUE TITLE
sci-mathematics/dikcube fails to build with GCC 14

### ISSUE THREAD

**Issue by joeyryan** [cschwan/sage-on-gentoo#797]
Due to GCC 14 making `implicit-int` and `implicit-function-declaration` errors by default, `sci-mathematics/dikcube` will fail to build on a newer Gentoo system that is using ~amd64 (and therefore, GCC 14)

Adding the following to my portage flags for this package allows this to build
CFLAGS="${CFLAGS} -Wno-error=implicit-function-declaration -Wno-error=implicit-int"
CXXFLAGS="${CXXFLAGS} -Wno-error=implicit-function-declaration -Wno-error=implicit-int"

Here is my build log:
```
>>> Emerging (1 of 1) sci-mathematics/dikcube-20070912-r1::sage-on-gentoo
>>> Failed to emerge sci-mathematics/dikcube-20070912-r1, Log file:
>>>  '/var/tmp/portage/sci-mathematics/dikcube-20070912-r1/temp/build.log'
>>> Jobs: 0 of 1 complete, 1 failed                                     Load avg: 0.74, 0.54, 0.46
 * Package:    sci-mathematics/dikcube-20070912-r1:0
 * Repository: sage-on-gentoo
 * Maintainer: frp.bissey@gmail.com
 * USE:        abi_x86_64 amd64 elibc_glibc kernel_linux
 * FEATURES:   network-sandbox preserve-libs sandbox userpriv usersandbox
>>> Unpacking source...
>>> Unpacking rubiks-20070912.tar.bz2 to /var/tmp/portage/sci-mathematics/dikcube-20070912-r1/work
>>> Source unpacked in /var/tmp/portage/sci-mathematics/dikcube-20070912-r1/work
>>> Preparing source in /var/tmp/portage/sci-mathematics/dikcube-20070912-r1/work/rubiks-20070912/dik ...
 * Applying dikcube-20070912_p10-fix-missing-includes.patch ...                                                                                                                                                           [ ok ]
 * Applying dikcube-20070912_p10-fix-LDFLAGS.patch ...                                                                                                                                                                    [ ok ]
>>> Source prepared.
>>> Configuring source in /var/tmp/portage/sci-mathematics/dikcube-20070912-r1/work/rubiks-20070912/dik ...
>>> Source configured.
>>> Compiling source in /var/tmp/portage/sci-mathematics/dikcube-20070912-r1/work/rubiks-20070912/dik ...
make -j14 CC=x86_64-pc-linux-gnu-gcc 'CFLAGS=-march=native -DLARGE_MEM -DVERBOSE' 'LDFLAGS=-Wl,-O1 -Wl,--as-needed -Wl,-z,pack-relative-relocs' all
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE   -c -o cube.o cube.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE   -c -o phase1.o phase1.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE   -c -o phase2.o phase2.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE   -c -o setcube.o setcube.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE   -c -o permcube.o permcube.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE   -c -o prntsol.o prntsol.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE   -c -o globals.o globals.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE   -c -o size222.o size222.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE -DCTWIST -c size333c.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE   -c -o sizesquare.o sizesquare.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE -DCCPERM -DIPERM -c sizedom.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE   -c -o sizekoc1.o sizekoc1.c
x86_64-pc-linux-gnu-gcc -march=native -DLARGE_MEM -DVERBOSE -DCCPERM -DIPERM -c sizekoc2.c
prntsol.c:5:1: error: return type defaults to ‘int’ [-Wimplicit-int]
    5 | prnt_sol()
      | ^~~~~~~~
permcube.c:3:1: error: return type defaults to ‘int’ [-Wimplicit-int]
    3 | perm_cube()
      | ^~~~~~~~~
size222.c:22:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   22 | do_pr(turns, count)
      | ^~~~~
size222.c:34:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   34 | try_one(num, count)
      | ^~~~~~~
size222.c:69:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   69 | try(count)
      | ^~~
size222.c:98:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   98 | main(argc, argv)
      | ^~~~
make: *** [<builtin>: prntsol.o] Error 1
make: *** Waiting for unfinished jobs....
make: *** [<builtin>: permcube.o] Error 1
make: *** [<builtin>: size222.o] Error 1
setcube.c:28:8: error: type defaults to ‘int’ in declaration of ‘cube_found’ [-Wimplicit-int]
   28 | static cube_found, orient_found;
      |        ^~~~~~~~~~
setcube.c:28:20: error: type defaults to ‘int’ in declaration of ‘orient_found’ [-Wimplicit-int]
   28 | static cube_found, orient_found;
      |                    ^~~~~~~~~~~~
cube.c:10:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   10 | main(argc, argv)
      | ^~~~
cube.c: In function ‘main’:
cube.c:52:5: error: implicit declaration of function ‘init_phase1’ [-Wimplicit-function-declaration]
   52 |     init_phase1();
      |     ^~~~~~~~~~~
phase2.c: In function ‘do_phase2’:
cube.c:54:5: error: implicit declaration of function ‘init_phase2’ [-Wimplicit-function-declaration]
   54 |     init_phase2();
      |     ^~~~~~~~~~~
phase2.c:107:9: error: implicit declaration of function ‘prnt_sol’ [-Wimplicit-function-declaration]
  107 |         prnt_sol();
      |         ^~~~~~~~
cube.c:65:9: error: implicit declaration of function ‘set_cube’; did you mean ‘def_cube’? [-Wimplicit-function-declaration]
   65 |         set_cube();
      |         ^~~~~~~~
      |         def_cube
make: *** [<builtin>: setcube.o] Error 1
phase1.c: In function ‘do_phase1’:
cube.c:69:17: error: implicit declaration of function ‘init_phase1_cube’ [-Wimplicit-function-declaration]
   69 |     do_phase1 = init_phase1_cube();
      |                 ^~~~~~~~~~~~~~~~
phase1.c:190:9: error: implicit declaration of function ‘perm_cube’ [-Wimplicit-function-declaration]
  190 |         perm_cube();
      |         ^~~~~~~~~
cube.c:76:13: error: implicit declaration of function ‘phase1’ [-Wimplicit-function-declaration]
   76 |             phase1(i++);
      |             ^~~~~~
phase1.c:191:21: error: implicit declaration of function ‘init_phase2_cube’; did you mean ‘init_phase1_cube’? [-Wimplicit-function-declaration]
  191 |         do_phase2 = init_phase2_cube();
      |                     ^~~~~~~~~~~~~~~~
      |                     init_phase1_cube
phase1.c:198:17: error: implicit declaration of function ‘phase2’ [-Wimplicit-function-declaration]
  198 |                 phase2(i++ - cur_move, dont1, dont2);
      |                 ^~~~~~
cube.c:82:9: error: implicit declaration of function ‘perm_cube’; did you mean ‘read_cube’? [-Wimplicit-function-declaration]
   82 |         perm_cube();
      |         ^~~~~~~~~
      |         read_cube
phase1.c:204:13: error: implicit declaration of function ‘prnt_sol’ [-Wimplicit-function-declaration]
  204 |             prnt_sol();
      |             ^~~~~~~~
cube.c:83:21: error: implicit declaration of function ‘init_phase2_cube’ [-Wimplicit-function-declaration]
   83 |         do_phase2 = init_phase2_cube();
      |                     ^~~~~~~~~~~~~~~~
cube.c:91:17: error: implicit declaration of function ‘phase2’ [-Wimplicit-function-declaration]
   91 |                 phase2(i++, -1, -1);
      |                 ^~~~~~
In file included from sizesquare.c:7:
trans/cperm2.a: In function ‘permcx_val’:
trans/cperm2.a:21:12: error: type of ‘p8’ defaults to ‘int’ [-Wimplicit-int]
   21 | static int permcx_val(p0,p1,p2,p3,p4,p5,p6,p7,p8)
      |            ^~~~~~~~~~
make: *** [<builtin>: phase2.o] Error 1
size333c.c:43:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   43 | do_pr(turns, count)
      | ^~~~~
size333c.c:55:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   55 | try_one(num, count)
      | ^~~~~~~
size333c.c:127:1: error: return type defaults to ‘int’ [-Wimplicit-int]
  127 | try(count)
      | ^~~
size333c.c:156:1: error: return type defaults to ‘int’ [-Wimplicit-int]
  156 | main(argc, argv)
      | ^~~~
sizesquare.c: At top level:
sizesquare.c:21:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   21 | do_pr(turns, count)
      | ^~~~~
sizesquare.c:33:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   33 | try_one(num, count)
      | ^~~~~~~
make: *** [<builtin>: phase1.o] Error 1
sizesquare.c:61:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   61 | try(count)
      | ^~~
sizesquare.c:90:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   90 | main(argc, argv)
      | ^~~~
sizekoc2.c:33:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   33 | do_pr(turns, count)
      | ^~~~~
sizekoc2.c:45:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   45 | try_one(num, count)
      | ^~~~~~~
sizekoc2.c:133:1: error: return type defaults to ‘int’ [-Wimplicit-int]
  133 | try(count)
      | ^~~
sizekoc1.c:33:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   33 | do_pr(turns, count)
      | ^~~~~
sizekoc2.c:162:1: error: return type defaults to ‘int’ [-Wimplicit-int]
  162 | main(argc, argv)
      | ^~~~
sizekoc1.c:45:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   45 | try_one(num, count)
      | ^~~~~~~
sizekoc1.c:96:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   96 | try(count)
      | ^~~
sizedom.c:44:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   44 | do_pr(turns, count)
      | ^~~~~
sizekoc1.c:125:1: error: return type defaults to ‘int’ [-Wimplicit-int]
  125 | main(argc, argv)
      | ^~~~
make: *** [<builtin>: cube.o] Error 1
sizedom.c:56:1: error: return type defaults to ‘int’ [-Wimplicit-int]
   56 | try_one(num, count)
      | ^~~~~~~
sizedom.c:180:1: error: return type defaults to ‘int’ [-Wimplicit-int]
  180 | try(count)
      | ^~~
sizedom.c:278:1: error: return type defaults to ‘int’ [-Wimplicit-int]
  278 | main(argc, argv)
      | ^~~~
make: *** [<builtin>: sizesquare.o] Error 1
make: *** [makefile:20: size333c.o] Error 1
make: *** [<builtin>: sizekoc1.o] Error 1
make: *** [makefile:42: sizekoc2.o] Error 1
make: *** [makefile:31: sizedom.o] Error 1
 * ERROR: sci-mathematics/dikcube-20070912-r1::sage-on-gentoo failed (compile phase):
 *   emake failed
 *
 * If you need support, post the output of `emerge --info '=sci-mathematics/dikcube-20070912-r1::sage-on-gentoo'`,
 * the complete build log and the output of `emerge -pqv '=sci-mathematics/dikcube-20070912-r1::sage-on-gentoo'`.
 * The complete build log is located at '/var/tmp/portage/sci-mathematics/dikcube-20070912-r1/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/sci-mathematics/dikcube-20070912-r1/temp/environment'.
 * Working directory: '/var/tmp/portage/sci-mathematics/dikcube-20070912-r1/work/rubiks-20070912/dik'
 * S: '/var/tmp/portage/sci-mathematics/dikcube-20070912-r1/work/rubiks-20070912/dik'

 * Messages for package sci-mathematics/dikcube-20070912-r1:

 * ERROR: sci-mathematics/dikcube-20070912-r1::sage-on-gentoo failed (compile phase):
 *   emake failed
 *
 * If you need support, post the output of `emerge --info '=sci-mathematics/dikcube-20070912-r1::sage-on-gentoo'`,
 * the complete build log and the output of `emerge -pqv '=sci-mathematics/dikcube-20070912-r1::sage-on-gentoo'`.
 * The complete build log is located at '/var/tmp/portage/sci-mathematics/dikcube-20070912-r1/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/sci-mathematics/dikcube-20070912-r1/temp/environment'.
 * Working directory: '/var/tmp/portage/sci-mathematics/dikcube-20070912-r1/work/rubiks-20070912/dik'
 * S: '/var/tmp/portage/sci-mathematics/dikcube-20070912-r1/work/rubiks-20070912/dik'
```

**Comment by kiwifb** [comment]
Thank you for the report. I am torn between three between 3 options in order of difficulty:

1. drop the package since it is already optional upstream. The fact that it is abandonware and only maintained in sage makes this a serious option.
2. Add the flags in the ebuild itself.
3. Fix the source.

**Comment by joeyryan** [comment]
Considering it's a small program and the errors aren't impacting functionality, I wouldn't be opposed to adding the flags to the ebuild.

The functions with no return type seem to not return anything (void), and the implicit function declarations are referencing the correct functions. It's just modern GCC (and Clang) being extra cautious.

Since it's abandonware, it seems the source code is not in version control, correct? I'm tempted to fix it myself as well since it would only take 5 minutes to correct these issues, but I'm not sure where to contribute the changes back to.

I would consider removing it as a last resort option if it was broken. The changes to GCC are pretty recent and it's been compiling fine for 17 years now, with minimal complaints. :) I'm not big into Rubik's cubes but I'm sure someone out there finds use in the package.

Happy to help!

**Comment by kiwifb** [comment]
Well, there are two other rubik's cube solving package available (also abandonware mind you): cubex and optimal (plus cu2 but it only does 2x2x2).

I think the path of least resistance for now is to add flags.

**Comment by kiwifb** [comment]
The overlay should be synchronised in about 15 minutes and ready to be tested on a real compiler.

**Comment by joeyryan** [comment]
It now compiles successfully. I removed the flags that I added locally and reinstalled it. Before I synced the repo, it failed to install, and then after syncing the repo, it installed successfully.  (Just wanted to be sure)

Thanks for fixing this. I think it's a good solution considering it's abandonware as you said, and it's a very small program that's highly unlikely to have any issues with the flags added.
